src/algorand_transactions/transactions_group.rs

```diff
@@ -0,0 +1,220 @@
+use serde::{Deserialize, Serialize, Serializer};
+use sha2::Digest;
+
+use crate::{
+    algorand_errors::AlgorandError,
+    algorand_types::{Byte, Bytes, Result},
+    AlgorandHash,
+    AlgorandSignedTransaction,
+    AlgorandTransaction,
+};
+
+#[derive(Debug, Clone, Eq, PartialEq, Deserialize)]
+pub struct TxGroup {
+    #[serde(rename = "txlist", default)]
+    tx_group_hashes: Vec<AlgorandHash>,
+}
+
+impl TxGroup {
+    const MAX_TX_GROUP_SIZE: usize = 16;
+
+    pub fn new(tx_group_hashes: Vec<AlgorandHash>) -> TxGroup {
+        TxGroup { tx_group_hashes }
+    }
+
+    pub fn assign_group_id(txns: &mut [&mut AlgorandTransaction]) -> Result<AlgorandHash> {
+        let gid = TxGroup::compute_group_id(txns)?;
+        for tx in txns {
+            tx.assign_group_id(gid);
+        }
+        Ok(gid)
+    }
+
+    pub(crate) fn compute_group_id(txns: &[&mut AlgorandTransaction]) -> Result<AlgorandHash> {
+        if txns.is_empty() {
+            return Err(AlgorandError::Custom("Empty transactions list".to_string()));
+        }
+        if txns.len() > Self::MAX_TX_GROUP_SIZE {
+            return Err(AlgorandError::Custom("Too much transaction".to_string()));
+        }
+        let mut ids: Vec<AlgorandHash> = vec![];
+        for t in txns {
+            ids.push(t.to_raw_tx_id()?);
+        }
+        let group = TxGroup::new(ids);
+        let hashed = sha2::Sha512_256::digest(&group.encode_for_signing()?);
+        AlgorandHash::from_slice(hashed.as_slice())
+    }
+
+    fn to_msg_pack_bytes(&self) -> Result<Bytes> {
+        Ok(rmp_serde::to_vec_named(&self)?)
+    }
+
+    fn prefix_group_byte(bytes: &[Byte]) -> Bytes {
+        let suffix = bytes;
+        let mut prefix = b"TG".to_vec();
+        prefix.extend_from_slice(suffix);
+        prefix
+    }
+
+    pub fn encode_for_signing(&self) -> Result<Bytes> {
+        self.to_msg_pack_bytes()
+            .map(|ref msg_pack_bytes| Self::prefix_group_byte(msg_pack_bytes))
+    }
+
+    pub(crate) fn concat_signed_transactions(txns: &[AlgorandSignedTransaction]) -> Result<String> {
```

Comment:
This should be `pub` not `pub(crate)`.

Also, this is the final hex for pushing to a node right? The user doesn't need to know or care that it's concatting the txs. So let's name it similarly to the fxn call we make to the underlying txs:

```suggestion
     pub fn to_hex(txns: &[AlgorandSignedTransaction]) -> Result<String> {
```

Reply:
Also, when re-architectured per my main review comment, this fxn signature will become:

```suggestion
    pub fn to_hex(&self) -> Result<String> {
```

Reply:
`to_hex` no longer exists after refactoring.